# Importing necessary libraries

In [1]:
import bs4 as bs
import urllib
import urllib.request

# Webscrapping from link

In [2]:
import requests
amazonurl = requests.get("https://www.amazon.in/Amazon-FireTV-Stick-4K-Alexa-Voice-Remote-Streaming-Player/product-reviews/B079QQZZJK/ref=cm_cr_dp_d_show_all_btm?ie=UTF8&reviewerType=all_reviews")
html_contents=amazonurl.text
html_contents

'<!doctype html><html lang="en-in" class="a-no-js" data-19ax5a9jf="dingo"><!-- sp:feature:head-start -->\n<head><script>var aPageStart = (new Date()).getTime();</script><meta charset="utf-8"/>\n\n<!-- sp:feature:cs-optimization -->\n<meta http-equiv=\'x-dns-prefetch-control\' content=\'on\'>\n<link rel="dns-prefetch" href="https://images-eu.ssl-images-amazon.com">\n<link rel="dns-prefetch" href="https://m.media-amazon.com">\n<link rel="dns-prefetch" href="https://completion.amazon.com">\n\n<!-- sp:feature:aui-assets -->\n<link rel="stylesheet" href="https://images-eu.ssl-images-amazon.com/images/I/11EIQ5IGqaL._RC|012LjolmrML.css,41E3lUjtoNL.css,11cMnOipjJL.css,017DsKjNQJL.css,01Vctty9pOL.css,01zAp95w-aL.css,41EWOOlBJ9L.css,11UoGyLuXoL.css,01ElnPiDxWL.css,11QxHU4QYaL.css,01Sp8sB1HiL.css,01IdKcBuAdL.css,01y-XAlI+2L.css,01evdoiemkL.css,01K+Ps1DeEL.css,314djKvMsUL.css,01ZTetsDh7L.css,01pbA9Lg3yL.css,21LK7jaicML.css,11L58Qpo0GL.css,21kyTi1FabL.css,01ruG+gDPFL.css,01YhS3Cs-hL.css,21GwE3cR-yL

# Html tags are removed using BeautifulSoup

In [3]:
soup=bs.BeautifulSoup(html_contents,'html.parser')

In [4]:
soup

<!DOCTYPE doctype html>
<html class="a-no-js" data-19ax5a9jf="dingo" lang="en-in"><!-- sp:feature:head-start -->
<head><script>var aPageStart = (new Date()).getTime();</script><meta charset="utf-8"/>
<!-- sp:feature:cs-optimization -->
<meta content="on" http-equiv="x-dns-prefetch-control"/>
<link href="https://images-eu.ssl-images-amazon.com" rel="dns-prefetch"/>
<link href="https://m.media-amazon.com" rel="dns-prefetch"/>
<link href="https://completion.amazon.com" rel="dns-prefetch"/>
<!-- sp:feature:aui-assets -->
<link href="https://images-eu.ssl-images-amazon.com/images/I/11EIQ5IGqaL._RC|012LjolmrML.css,41E3lUjtoNL.css,11cMnOipjJL.css,017DsKjNQJL.css,01Vctty9pOL.css,01zAp95w-aL.css,41EWOOlBJ9L.css,11UoGyLuXoL.css,01ElnPiDxWL.css,11QxHU4QYaL.css,01Sp8sB1HiL.css,01IdKcBuAdL.css,01y-XAlI+2L.css,01evdoiemkL.css,01K+Ps1DeEL.css,314djKvMsUL.css,01ZTetsDh7L.css,01pbA9Lg3yL.css,21LK7jaicML.css,11L58Qpo0GL.css,21kyTi1FabL.css,01ruG+gDPFL.css,01YhS3Cs-hL.css,21GwE3cR-yL.css,11KLBtpWIAL.css,

In [5]:
print(soup.prettify())

<!DOCTYPE doctype html>
<html class="a-no-js" data-19ax5a9jf="dingo" lang="en-in">
 <!-- sp:feature:head-start -->
 <head>
  <script>
   var aPageStart = (new Date()).getTime();
  </script>
  <meta charset="utf-8"/>
  <!-- sp:feature:cs-optimization -->
  <meta content="on" http-equiv="x-dns-prefetch-control"/>
  <link href="https://images-eu.ssl-images-amazon.com" rel="dns-prefetch"/>
  <link href="https://m.media-amazon.com" rel="dns-prefetch"/>
  <link href="https://completion.amazon.com" rel="dns-prefetch"/>
  <!-- sp:feature:aui-assets -->
  <link href="https://images-eu.ssl-images-amazon.com/images/I/11EIQ5IGqaL._RC|012LjolmrML.css,41E3lUjtoNL.css,11cMnOipjJL.css,017DsKjNQJL.css,01Vctty9pOL.css,01zAp95w-aL.css,41EWOOlBJ9L.css,11UoGyLuXoL.css,01ElnPiDxWL.css,11QxHU4QYaL.css,01Sp8sB1HiL.css,01IdKcBuAdL.css,01y-XAlI+2L.css,01evdoiemkL.css,01K+Ps1DeEL.css,314djKvMsUL.css,01ZTetsDh7L.css,01pbA9Lg3yL.css,21LK7jaicML.css,11L58Qpo0GL.css,21kyTi1FabL.css,01ruG+gDPFL.css,01YhS3Cs-hL.css,21

In [6]:
import datetime
now=datetime.datetime.now()
print(soup.title.text,now)


Amazon.in:Customer reviews: Fire TV Stick 4K with Alexa Voice Remote | Stream in 4K resolution 2020-11-10 13:20:33.333459


# Getting the reviews

In [7]:
a=soup.findAll('span',{'class':"a-size-base review-text review-text-content"})
a

[<span class="a-size-base review-text review-text-content" data-hook="review-body">
 <span>
   I have a 🔥TV (non 4K) which I found to be excellent and when the 4K version was stated for release,  I pre ordered it.  However,  I am extremely disappointed with it.  I have a 55 inch Android VU TV,  2018 model which supports 4k as evidenced by 4k videos from YouTube but for the fire stick 4k,  the brightness and contrast was pre set to a dismal low value.  It was too dark to watch.  There is nothing in it to adjust my brightness and I gave up and went back to my fire TV non 4k one which atleast looks vibrant and joyful to watch. The 4k one is just appalling.  I am sorry but Amazon totally let me down here.
 </span>
 </span>,
 <span class="a-size-base review-text review-text-content" data-hook="review-body">
 <span>
   I have old firestick, bought this new 4K stick.. to be frank i don't see any difference eventhough its showing as playing Ultra HD there is no much difference for 1080p and Ul

In [8]:
lines=[s.get_text() for s in a]
lines

['\n\n  I have a 🔥TV (non 4K) which I found to be excellent and when the 4K version was stated for release,  I pre ordered it.  However,  I am extremely disappointed with it.  I have a 55 inch Android VU TV,  2018 model which supports 4k as evidenced by 4k videos from YouTube but for the fire stick 4k,  the brightness and contrast was pre set to a dismal low value.  It was too dark to watch.  There is nothing in it to adjust my brightness and I gave up and went back to my fire TV non 4k one which atleast looks vibrant and joyful to watch. The 4k one is just appalling.  I am sorry but Amazon totally let me down here.\n\n',
 "\n\n  I have old firestick, bought this new 4K stick.. to be frank i don't see any difference eventhough its showing as playing Ultra HD there is no much difference for 1080p and Ultra HD. Sound quality in my sound bar is okay... If you really don't care about 4K can still buy the old stick..I returned it to save my 6k. I'm 1000% satisfied with my old one.\n\n",
 '\

In [9]:
type(lines)

list

# Using Sentence Tokenization and Word Tokenization

In [10]:
import nltk
from nltk.tokenize import word_tokenize
from nltk.tokenize import sent_tokenize
nltk.download('punkt')
from nltk.corpus import stopwords
nltk.download('stopwords')
sentence=nltk.sent_tokenize(str(lines))
sentence
    


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


["['\\n\\n  I have a 🔥TV (non 4K) which I found to be excellent and when the 4K version was stated for release,  I pre ordered it.",
 'However,  I am extremely disappointed with it.',
 'I have a 55 inch Android VU TV,  2018 model which supports 4k as evidenced by 4k videos from YouTube but for the fire stick 4k,  the brightness and contrast was pre set to a dismal low value.',
 'It was too dark to watch.',
 'There is nothing in it to adjust my brightness and I gave up and went back to my fire TV non 4k one which atleast looks vibrant and joyful to watch.',
 'The 4k one is just appalling.',
 'I am sorry but Amazon totally let me down here.\\n\\n\', "\\n\\n  I have old firestick, bought this new 4K stick.. to be frank i don\'t see any difference eventhough its showing as playing Ultra HD there is no much difference for 1080p and Ultra HD.',
 'Sound quality in my sound bar is okay...',
 "If you really don't care about 4K can still buy the old stick..I returned it to save my 6k.",
 'I\'m 1

In [11]:
words=nltk.word_tokenize(str(sentence))
len(words)

6165

# Data Cleaning

In [12]:
import re
for i in range(len(sentence)):
    sentence[i]=sentence[i].lower()
    sentence[i]=re.sub(r'\W',' ',str(sentence[i]))
    sentence[i]=re.sub(r'\s+',' ',str(sentence[i]))
    sentence[i]=sentence[i].replace('n n',' ')
    sentence[i]=sentence[i].replace("'",' ')
sentence



['   i have a tv non 4k which i found to be excellent and when the 4k version was stated for release i pre ordered it ',
 'however i am extremely disappointed with it ',
 'i have a 55 inch android vu tv 2018 model which supports 4k as evidenced by 4k videos from youtube but for the fire stick 4k the brightness and contrast was pre set to a dismal low value ',
 'it was too dark to watch ',
 'there is nothing in it to adjust my brightness and i gave up and went back to my fire tv non 4k one which atleast looks vibrant and joyful to watch ',
 'the 4k one is just appalling ',
 'i am sorry but amazon totally let me down here     i have old firestick bought this new 4k stick to be frank i don t see any difference eventhough its showing as playing ultra hd there is no much difference for 1080p and ultra hd ',
 'sound quality in my sound bar is okay ',
 'if you really don t care about 4k can still buy the old stick i returned it to save my 6k ',
 'i m 1000 satisfied with my old one     firstly

In [13]:
len(sentence)

210

# Removing StopWords

In [14]:
for i in range(len(sentence)):
    words=nltk.word_tokenize(sentence[i])
    stopwords_removed=[word for word in words if word not in stopwords.words('english')]
    sentence[i]=' '.join(stopwords_removed)
len(stopwords.words('english'))

179

In [15]:
sentence

['tv non 4k found excellent 4k version stated release pre ordered',
 'however extremely disappointed',
 '55 inch android vu tv 2018 model supports 4k evidenced 4k videos youtube fire stick 4k brightness contrast pre set dismal low value',
 'dark watch',
 'nothing adjust brightness gave went back fire tv non 4k one atleast looks vibrant joyful watch',
 '4k one appalling',
 'sorry amazon totally let old firestick bought new 4k stick frank see difference eventhough showing playing ultra hd much difference 1080p ultra hd',
 'sound quality sound bar okay',
 'really care 4k still buy old stick returned save 6k',
 '1000 satisfied old one firstly early adopter technology bought amazon fire stick launched currently using 7 tvs well car simply love freedom watching favourite entertainment anywhere everywhere excited see amazon launching 4k fire stick due course time tvs upgraded 4k however li l content watch uhd ordered 4k fire stick last month announced ordering delivery long wait took month am

In [16]:
words=nltk.word_tokenize(str(sentence))
len(words)

3640

# Getting frequency of each word

In [17]:
from collections import Counter
w=[i for i in words]
count=Counter(w)
count

Counter({'[': 1,
         "'tv": 1,
         'non': 3,
         '4k': 47,
         'found': 3,
         'excellent': 1,
         'version': 1,
         'stated': 1,
         'release': 1,
         'pre': 3,
         'ordered': 2,
         "'": 218,
         ',': 209,
         "'however": 1,
         'extremely': 3,
         'disappointed': 4,
         "'55": 1,
         'inch': 1,
         'android': 10,
         'vu': 3,
         'tv': 96,
         '2018': 1,
         'model': 1,
         'supports': 4,
         'evidenced': 1,
         'videos': 1,
         'youtube': 6,
         'fire': 48,
         'stick': 40,
         'brightness': 2,
         'contrast': 1,
         'set': 3,
         'dismal': 1,
         'low': 3,
         'value': 1,
         "'dark": 1,
         'watch': 7,
         "'nothing": 1,
         'adjust': 2,
         'gave': 1,
         'went': 1,
         'back': 2,
         'one': 19,
         'atleast': 1,
         'looks': 1,
         'vibrant': 1,
         'j

In [18]:
count.most_common(20)

[("'", 218),
 (',', 209),
 ('tv', 96),
 ('fire', 48),
 ('4k', 47),
 ('stick', 40),
 ('content', 37),
 ('apps', 30),
 ('remote', 29),
 ('device', 29),
 ('sec', 26),
 ('amazon', 24),
 ('app', 24),
 ('screen', 22),
 ('via', 21),
 ('1', 21),
 ('smart', 20),
 ('install', 20),
 ('one', 19),
 ('5', 19)]

# Lematisation

In [19]:
from nltk.stem import WordNetLemmatizer
lematizer=WordNetLemmatizer()
for i in range(len(sentence)):
    words=nltk.word_tokenize(sentence[i])
    new=[lematizer.lemmatize(word) for word in words]
    sentence[i]=' '.join(new)
sentence

['tv non 4k found excellent 4k version stated release pre ordered',
 'however extremely disappointed',
 '55 inch android vu tv 2018 model support 4k evidenced 4k video youtube fire stick 4k brightness contrast pre set dismal low value',
 'dark watch',
 'nothing adjust brightness gave went back fire tv non 4k one atleast look vibrant joyful watch',
 '4k one appalling',
 'sorry amazon totally let old firestick bought new 4k stick frank see difference eventhough showing playing ultra hd much difference 1080p ultra hd',
 'sound quality sound bar okay',
 'really care 4k still buy old stick returned save 6k',
 '1000 satisfied old one firstly early adopter technology bought amazon fire stick launched currently using 7 tv well car simply love freedom watching favourite entertainment anywhere everywhere excited see amazon launching 4k fire stick due course time tv upgraded 4k however li l content watch uhd ordered 4k fire stick last month announced ordering delivery long wait took month amazon 

# Visualisation using WordCloud

In [20]:
from wordcloud import WordCloud
import heapq
import matplotlib.pyplot as plt
wc=WordCloud(width=400,height=300,max_words=100,background_color='white').generate_from_frequencies(count)
plt.rcParams['figure.figsize'] = (8, 10)
plt.imshow(wc,interpolation='bilinear')
plt.axis('off')
plt.show()

In C:\ProgramData\Anaconda3\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle: 
The text.latex.preview rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.
In C:\ProgramData\Anaconda3\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle: 
The mathtext.fallback_to_cm rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.
In C:\ProgramData\Anaconda3\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle: Support for setting the 'mathtext.fallback_to_cm' rcParam is deprecated since 3.3 and will be removed two minor releases later; use 'mathtext.fallback : 'cm' instead.
In C:\ProgramData\Anaconda3\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle: 
The validate_bool_maybe_none function was deprecated in Matplotlib 3.3 and will be removed two minor releases later.
In C:\ProgramData\Anaconda3\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mpl

<Figure size 800x1000 with 1 Axes>

# Using Pos_tags

In [21]:
tagged_words=nltk.pos_tag(words)
word_tags=[]
for tw in tagged_words:
    word_tags.append(tw[0]+'-'+tw[1])
word_tags

['many-JJ', 'work-NN', 'software-NN', 'upgrade-JJ', 'done-VBN']

# Getting top 200 words

In [22]:
freq_words=heapq.nlargest(200,count,key=count.get)
freq_words

["'",
 ',',
 'tv',
 'fire',
 '4k',
 'stick',
 'content',
 'apps',
 'remote',
 'device',
 'sec',
 'amazon',
 'app',
 'screen',
 'via',
 '1',
 'smart',
 'install',
 'one',
 '5',
 'home',
 'launch',
 'volume',
 '2',
 'even',
 'available',
 'control',
 'prime',
 '3',
 'etc',
 'firestick',
 'quality',
 '4',
 'netflix',
 'alexa',
 'firetv',
 'button',
 'also',
 'hdmi',
 'music',
 '7',
 'time',
 'hdr',
 'voice',
 'cast',
 'video',
 'android',
 'previous',
 'better',
 'like',
 'good',
 'support',
 'settings',
 '1080p',
 'well',
 'performance',
 'update',
 'mobile',
 '6',
 'atmos',
 'difference',
 'much',
 'switch',
 'devices',
 'may',
 'streaming',
 'mirroring',
 'ir',
 'cec',
 'watch',
 "'4k",
 'old',
 'hd',
 'using',
 'uhd',
 'make',
 'work',
 'feature',
 'subscription',
 'works',
 'free',
 'g',
 '8',
 'another',
 'e',
 'youtube',
 'sound',
 'however',
 'power',
 'high',
 'would',
 'normal',
 'times',
 'hotstar',
 'bt',
 'cable',
 'mm',
 'fi',
 'required',
 'live',
 'es',
 'vs',
 'see',
 'st

# Finding inverse document frequency of all sentences

In [23]:
import pandas as pd
import numpy as np
idf_matrix={}
for word in freq_words:
    doc_cnt=0
    for data in sentence:
        if word in nltk.word_tokenize(data):
            doc_cnt+=1
        try:
            z=((len(sentence)/doc_cnt)+1)
        except ZeroDivisionError:
            z = 0
            
    idf_matrix[word]=np.log(z+1)
idf_matrix

{"'": 0.0,
 ',': 0.0,
 'tv': 1.7047480922384253,
 'fire': 2.0380563255169815,
 '4k': 2.0380563255169815,
 'stick': 2.1473922035883435,
 'content': 2.2512917986064953,
 'apps': 2.3749057545736716,
 'remote': 2.2512917986064953,
 'device': 2.1972245773362196,
 'sec': 4.276666119016055,
 'amazon': 2.3749057545736716,
 'app': 2.4462918136602205,
 'screen': 2.833213344056216,
 'via': 2.663954881236986,
 '1': 2.716349003916905,
 'smart': 3.048842098867939,
 'install': 3.4657359027997265,
 'one': 2.663954881236986,
 '5': 2.8988824475640733,
 'home': 2.772588722239781,
 'launch': 3.6109179126442243,
 'volume': 3.048842098867939,
 '2': 2.833213344056216,
 'even': 3.1354942159291497,
 'available': 2.970414465569701,
 'control': 2.8988824475640733,
 'prime': 2.833213344056216,
 '3': 2.833213344056216,
 'etc': 2.970414465569701,
 'firestick': 3.048842098867939,
 'quality': 2.970414465569701,
 '4': 2.970414465569701,
 'netflix': 3.048842098867939,
 'alexa': 2.772588722239781,
 'firetv': 3.048842098

# Finding term frequency of all sentences

In [24]:
tf_matrix={}
for word in freq_words:
    doc_tf=[]
    for data in sentence:
        frequency=0
        if word in nltk.word_tokenize(data):
            frequency+=1
        tf_word=frequency/len(nltk.word_tokenize(data))
        doc_tf.append(tf_word)
    tf_matrix[word]=doc_tf
tf_matrix

{"'": [0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
 

# Finding tfidf score for all words in sentence

In [25]:
tfidfmatrix=[]
for word in tf_matrix.keys():
    tfidf=[]
    for value in tf_matrix[word]:
        score=value*idf_matrix[word]
        tfidf.append(score)
    tfidfmatrix.append(tfidf)
tfidfmatrix

[[0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,
  0.0,

# Bag of words

In [26]:
wordnet={}
for i in sentence:
    words=nltk.word_tokenize(i)
    for word in words:
        if word not in wordnet.keys():
            wordnet[word]=1
        else:
            wordnet[word]+=1


In [27]:
wordnet

{'tv': 100,
 'non': 3,
 '4k': 54,
 'found': 3,
 'excellent': 1,
 'version': 1,
 'stated': 1,
 'release': 1,
 'pre': 3,
 'ordered': 2,
 'however': 7,
 'extremely': 3,
 'disappointed': 5,
 '55': 5,
 'inch': 3,
 'android': 10,
 'vu': 3,
 '2018': 1,
 'model': 3,
 'support': 15,
 'evidenced': 1,
 'video': 13,
 'youtube': 8,
 'fire': 50,
 'stick': 44,
 'brightness': 2,
 'contrast': 1,
 'set': 7,
 'dismal': 1,
 'low': 3,
 'value': 3,
 'dark': 1,
 'watch': 8,
 'nothing': 1,
 'adjust': 2,
 'gave': 1,
 'went': 1,
 'back': 2,
 'one': 20,
 'atleast': 1,
 'look': 3,
 'vibrant': 1,
 'joyful': 1,
 'appalling': 1,
 'sorry': 1,
 'amazon': 30,
 'totally': 2,
 'let': 4,
 'old': 7,
 'firestick': 15,
 'bought': 4,
 'new': 4,
 'frank': 1,
 'see': 5,
 'difference': 10,
 'eventhough': 1,
 'showing': 1,
 'playing': 3,
 'ultra': 2,
 'hd': 7,
 'much': 10,
 '1080p': 10,
 'sound': 8,
 'quality': 14,
 'bar': 1,
 'okay': 1,
 'really': 3,
 'care': 2,
 'still': 5,
 'buy': 4,
 'returned': 2,
 'save': 2,
 '6k': 1,
 '100

In [28]:
x=[]
for i in sentence:
    words=nltk.word_tokenize(i)
    vector=[]
    for word in words:
        if word in freq_words:
            vector.append(1)
        else:
            vector.append(0)
    x.append(vector)

In [29]:
x

[[1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0],
 [1, 1, 1],
 [1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0],
 [0, 1],
 [0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1],
 [1, 1, 0],
 [0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1],
 [1, 1, 1, 0, 0],
 [0, 0, 1, 1, 1, 1, 1, 0, 0, 0],
 [0,
  0,
  1,
  1,
  0,
  0,
  0,
  0,
  1,
  1,
  1,
  1,
  1,
  0,
  1,
  1,
  1,
  1,
  0,
  1,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  1,
  1,
  0,
  1,
  1,
  1,
  0,
  0,
  1,
  1,
  0,
  1,
  1,
  0,
  0,
  1,
  1,
  1,
  0,
  1,
  1,
  1,
  0,
  0,
  0,
  0,
  0,
  1,
  1,
  0,
  0,
  1,
  0,
  1,
  1,
  1,
  0,
  0,
  1,
  1,
  1,
  1,
  1,
  0,
  0,
  1,
  1,
  0,
  0,
  0],
 [0,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  1,
  0,
  0,
  1,
  1,
  1,
  0,
  0,
  0,
  0,
  1,
  1,
  1,
  0,
  1,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
  0,
  1,
  1],
 [0, 0],
 [1,
  0,
  1,
  0,
  1,
  0,
  1,
  1,
  1,
  1,
  0,
  1,
  1,
  0,
  1,
  0,
  1,
  1,
  1,
  1,
  1,
  1,
  1,
  1,

# If word present in freq_word list then 1 is returned  else 0 is returned and thus we converted binary set of values representing bag of words model

In [30]:
import numpy as np
x=np.asarray(x)
x

array([list([1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0]), list([1, 1, 1]),
       list([1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0]),
       list([0, 1]),
       list([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1]),
       list([1, 1, 0]),
       list([0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1]),
       list([1, 1, 1, 0, 0]), list([0, 0, 1, 1, 1, 1, 1, 0, 0, 0]),
       list([0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0]),
       list([0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1]),
       list([0, 0]),
       list([1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0]),
       list([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1,

In [31]:
from sklearn.feature_extraction.text import CountVectorizer
vectorizer=CountVectorizer(max_features=500,min_df=2,max_df=0.5,stop_words=stopwords.words('english'))
X=vectorizer.fit_transform(sentence).toarray()
X

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 1, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

# Using TextBlob for sentiment prediction

In [32]:
from textblob import TextBlob


In [33]:
sentiment=[]
for i in sentence:
    text=TextBlob(i)
    blob=text.sentiment[0]
    sentiment.append(blob)
sentiment   
    

[1.0,
 -0.75,
 0.0,
 -0.15,
 0.08333333333333333,
 -0.35,
 -0.012727272727272731,
 0.43333333333333335,
 0.15000000000000002,
 0.06666666666666667,
 0.0047619047619047675,
 0.125,
 0.14166666666666666,
 0.07040816326530612,
 0.8,
 0.2,
 0.1776190476190476,
 -0.06666666666666667,
 -0.06574074074074074,
 0.15555555555555556,
 -0.4,
 0.20436507936507936,
 0.2833333333333334,
 -0.2916666666666667,
 0.7000000000000001,
 0.45000000000000007,
 0.02857142857142858,
 -0.06190476190476191,
 0.5062500000000001,
 0.26875,
 0.05000000000000002,
 0.22288888888888891,
 0.0,
 0.10476190476190476,
 0.5666666666666667,
 0.0,
 -0.1,
 0.0,
 0.0,
 0.0,
 0.3333333333333333,
 0.0,
 0.21000000000000002,
 0.3322222222222222,
 0.2,
 0.05,
 0.0,
 -0.024999999999999994,
 0.0,
 0.4,
 -0.05,
 0.2,
 0.3875,
 0.0,
 0.0,
 0.35,
 0.0,
 0.0,
 0.0,
 0.0,
 0.4,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.4166666666666667,
 0.0,
 0.057142857142857134,
 0.0,
 0.21428571428571427,
 -0.05,
 -0.1,
 0.5,
 0.08,
 0.0,
 0.

In [34]:
senti=[]
for i in sentiment:
    if i==0:
        senti.append('0')
    elif i>=1:
        senti.append('1')
    else:
        senti.append('-1')
  

In [35]:
senti

['1',
 '-1',
 '0',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '0',
 '-1',
 '-1',
 '0',
 '-1',
 '0',
 '0',
 '0',
 '-1',
 '0',
 '-1',
 '-1',
 '-1',
 '-1',
 '0',
 '-1',
 '0',
 '-1',
 '-1',
 '-1',
 '-1',
 '0',
 '0',
 '-1',
 '0',
 '0',
 '0',
 '0',
 '-1',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '0',
 '-1',
 '0',
 '-1',
 '0',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '0',
 '0',
 '0',
 '0',
 '-1',
 '0',
 '-1',
 '-1',
 '0',
 '0',
 '0',
 '0',
 '-1',
 '0',
 '0',
 '-1',
 '-1',
 '-1',
 '0',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '0',
 '-1',
 '-1',
 '-1',
 '0',
 '0',
 '0',
 '-1',
 '-1',
 '0',
 '0',
 '0',
 '-1',
 '-1',
 '0',
 '0',
 '-1',
 '0',
 '0',
 '0',
 '0',
 '-1',
 '0',
 '0',
 '-1',
 '-1',
 '0',
 '-1',
 '0',
 '0',
 '0',
 '0',
 '-1',
 '0',
 '-1',
 '0',
 '-1',
 '0',
 '-1',
 '0',
 '0',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',

# Creating a DataFrame using senti

In [36]:
sentiment_class=pd.DataFrame(senti,columns=['sentiment_class'])

In [37]:
sentence

['tv non 4k found excellent 4k version stated release pre ordered',
 'however extremely disappointed',
 '55 inch android vu tv 2018 model support 4k evidenced 4k video youtube fire stick 4k brightness contrast pre set dismal low value',
 'dark watch',
 'nothing adjust brightness gave went back fire tv non 4k one atleast look vibrant joyful watch',
 '4k one appalling',
 'sorry amazon totally let old firestick bought new 4k stick frank see difference eventhough showing playing ultra hd much difference 1080p ultra hd',
 'sound quality sound bar okay',
 'really care 4k still buy old stick returned save 6k',
 '1000 satisfied old one firstly early adopter technology bought amazon fire stick launched currently using 7 tv well car simply love freedom watching favourite entertainment anywhere everywhere excited see amazon launching 4k fire stick due course time tv upgraded 4k however li l content watch uhd ordered 4k fire stick last month announced ordering delivery long wait took month amazon 

In [38]:
reviews=pd.DataFrame(sentence,columns=['sentence'])
reviews

sentence
0    tv non 4k found excellent 4k version stated re...
1                       however extremely disappointed
2    55 inch android vu tv 2018 model support 4k ev...
3                                           dark watch
4    nothing adjust brightness gave went back fire ...
..                                                 ...
205  disappointed know netflix play atmos title 5 1...
206  practically atmos feature usable except watchi...
207                              like tv support hdr10
208  firestick hdr feature order make work switch a...
209                    many work software upgrade done

[210 rows x 1 columns]

In [39]:
type(sentiment_class)
type(sentence)

list

# Created a dataframe which contains sentences and their sentiment class

In [40]:
amazon_reviews=pd.concat([reviews,sentiment_class],axis=1)
amazon_reviews

sentence sentiment_class
0    tv non 4k found excellent 4k version stated re...               1
1                       however extremely disappointed              -1
2    55 inch android vu tv 2018 model support 4k ev...               0
3                                           dark watch              -1
4    nothing adjust brightness gave went back fire ...              -1
..                                                 ...             ...
205  disappointed know netflix play atmos title 5 1...              -1
206  practically atmos feature usable except watchi...              -1
207                              like tv support hdr10               0
208  firestick hdr feature order make work switch a...              -1
209                    many work software upgrade done              -1

[210 rows x 2 columns]

# Using countvectorizer to convert the sentences into binary bag of words model

In [41]:
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(max_features=1000,min_df=2,max_df=0.6,stop_words=stopwords.words('english'))
vector=cv.fit_transform(sentence).toarray()

In [42]:
vector

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 1, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

# Converting into TfidfTransformer

In [43]:
from sklearn.feature_extraction.text import TfidfTransformer
tfidf=TfidfTransformer()
tfidf_vector=tfidf.fit_transform(vector).toarray()

In [44]:
tfidf_vector

array([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.22922568,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

# Spliting the dataset into training and testing sets

In [45]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(tfidf_vector,senti,random_state=0,test_size=0.2)

# Using LogisticRegression

In [46]:
from sklearn.linear_model import LogisticRegression
lr=LogisticRegression()
trained=lr.fit(x_train,y_train)


In [47]:
y_pred=trained.predict(x_test)

In [48]:
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(y_test,y_pred)
cm

array([[22,  1],
       [12,  7]], dtype=int64)

# Out of 210 we got 13 datapoints misclassified

In [49]:
from sklearn.metrics import accuracy_score
print("Accuracy using Logistic Regression :",accuracy_score(y_test,y_pred)*100,'%')

Accuracy using Logistic Regression : 69.04761904761905 %


In [50]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

          -1       0.65      0.96      0.77        23
           0       0.88      0.37      0.52        19

    accuracy                           0.69        42
   macro avg       0.76      0.66      0.65        42
weighted avg       0.75      0.69      0.66        42



# By using Logistic Regression model is giving accuracy of 69%

In [51]:
from sklearn.model_selection import train_test_split
x_train_NB,x_test_NB,y_train_NB,y_test_NB=train_test_split(tfidf_vector,senti,random_state=0,test_size=0.2)

# Using Naive Bayes Classification

In [52]:
from sklearn.naive_bayes import GaussianNB
model=GaussianNB()
NB=model.fit(x_train_NB,y_train_NB)
NB_result=NB.predict(x_test_NB)

In [53]:
from sklearn.metrics import confusion_matrix
cm1=confusion_matrix(y_test_NB,NB_result)
cm1

array([[22,  1],
       [10,  9]], dtype=int64)

# When we use GaussianNB classifier out of 210 datapoints we got 11 datapoints are misclassified 

In [54]:
from sklearn.metrics import accuracy_score
print("Accuracy using Naive Bayes classification :",accuracy_score(y_test_NB,NB_result)*100,'%')

Accuracy using Naive Bayes classification : 73.80952380952381 %


In [55]:
from sklearn.metrics import classification_report
print(classification_report(y_test_NB,NB_result))

              precision    recall  f1-score   support

          -1       0.69      0.96      0.80        23
           0       0.90      0.47      0.62        19

    accuracy                           0.74        42
   macro avg       0.79      0.72      0.71        42
weighted avg       0.78      0.74      0.72        42



# By using Naive Bayes classification our model gives an accuracy of 74% 

In [56]:
import numpy as np
from sklearn.model_selection import train_test_split
x_train_NB,x_test_NB,y_train_NB,y_test_NB=train_test_split(tfidf_vector,senti,random_state=0,test_size=0.2)

# Using Multinomial NB

In [57]:
from sklearn.naive_bayes import MultinomialNB
MNB=MultinomialNB()
MNB_result=MNB.fit(x_train_NB,y_train_NB)

In [58]:
MNB_result = MNB.predict(x_test_NB)

In [59]:
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix
print("Accuracy using Multinomial NB:",accuracy_score(y_test_NB,MNB_result)*100,'%')

Accuracy using Multinomial NB: 61.904761904761905 %


In [60]:
from sklearn.metrics import confusion_matrix
cm2=confusion_matrix(y_test_NB,MNB_result)
cm2

array([[21,  2],
       [14,  5]], dtype=int64)

# Out of 210 datapoints we got 16 datapoints as misclassified

In [61]:
print(classification_report(y_test_NB,MNB_result))

              precision    recall  f1-score   support

          -1       0.60      0.91      0.72        23
           0       0.71      0.26      0.38        19

    accuracy                           0.62        42
   macro avg       0.66      0.59      0.55        42
weighted avg       0.65      0.62      0.57        42



# By using Multinomial Naive Bayes we get accuracy of 62%

In [62]:
from sklearn.model_selection import train_test_split
x_train_svm,x_test_svm,y_train_svm,y_test_svm=train_test_split(tfidf_vector,senti,random_state=0,test_size=0.2)

# Using Support Vector Machine

In [63]:
from sklearn.svm import SVC
svm=SVC(gamma='auto')
svm_result=svm.fit(x_train_svm,y_train_svm)
svm_result=svm_result.predict(x_test_svm)

In [64]:
from sklearn.metrics import accuracy_score
print("Accuracy using support vector machines:",accuracy_score(y_test_svm,svm_result)*100,'%')

Accuracy using support vector machines: 54.761904761904766 %


In [65]:
from sklearn.metrics import classification_report
print(classification_report(y_test_svm,svm_result))

              precision    recall  f1-score   support

          -1       0.55      1.00      0.71        23
           0       0.00      0.00      0.00        19

    accuracy                           0.55        42
   macro avg       0.27      0.50      0.35        42
weighted avg       0.30      0.55      0.39        42



C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [67]:
from sklearn.metrics import confusion_matrix
cm4=confusion_matrix(y_test_svm,svm_result)
cm4

array([[23,  0],
       [19,  0]], dtype=int64)

# Out of 210 datapoints we got 19 datapoints as misclassified and by using support vector machine we get accuracy of 55%

# Hence by doing webscrapping of reviews from amazon we found better accuracy was shown by using Naive Bayes Gaussian classifier we achieved accuracy upto 74%